src/main/java/net/smartcosmos/cluster/userdetails/impl/UserDetailsPersistenceService.java

```diff
@@ -1,49 +1,61 @@
 package net.smartcosmos.cluster.userdetails.impl;
 
-import java.util.LinkedHashSet;
 import java.util.Optional;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import lombok.extern.slf4j.Slf4j;
 
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.authentication.InternalAuthenticationServiceException;
+import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
 
 import net.smartcosmos.cluster.userdetails.dao.UserDetailsDao;
-import net.smartcosmos.cluster.userdetails.domain.AuthorityEntity;
 import net.smartcosmos.cluster.userdetails.domain.UserEntity;
 import net.smartcosmos.cluster.userdetails.dto.UserDetailsResponse;
 import net.smartcosmos.cluster.userdetails.repository.UserRepository;
 import net.smartcosmos.cluster.userdetails.util.UuidUtil;
 
-import static java.util.stream.Collectors.toSet;
-
 @Slf4j
 @Service
 public class UserDetailsPersistenceService implements UserDetailsDao {
 
     private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
 
     @Autowired
-    public UserDetailsPersistenceService(UserRepository userRepository) {
+    public UserDetailsPersistenceService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
 
         this.userRepository = userRepository;
+        this.passwordEncoder = passwordEncoder;
     }
 
     @Override
     public Optional<UserDetailsResponse> getAuthorities(String username, String password) {
 
-        Optional<UserEntity> userOptional = userRepository.getUserByCredentials(username, password);
+        if (StringUtils.isEmpty(password)) {
+            throw new InternalAuthenticationServiceException("Password blank.");
+        }
+
+        Optional<UserEntity> userOptional = userRepository.findByUsernameIgnoreCase(username);
         if (!userOptional.isPresent()) {
-            return Optional.empty();
+            throw new InternalAuthenticationServiceException("Unknown username " + username);
         }
 
         UserEntity user = userOptional.get();
-        Optional<Set<AuthorityEntity>> authorityOptional = userRepository.getAuthorities(user.getTenantId(), user.getId());
-        Set<AuthorityEntity> authorityEntities = authorityOptional.isPresent() ? authorityOptional.get() : new LinkedHashSet<>();
-        Set<String> authorities = authorityEntities.parallelStream()
-            .map(AuthorityEntity::getAuthority)
-            .collect(toSet());
+        if (!passwordEncoder.matches(password, user.getPassword())) {
+            throw new InternalAuthenticationServiceException("Password invalid for username " + username);
```

Comment:
the current accepted way is to not indicate the specific cause of failure, just something generic like "userid password credentials do not match"